### ISSUE TITLE
Search - JsonSerializationException when using model class with non-nullable value types

### ISSUE THREAD

**Issue by brjohnstmsft** [Azure/azure-sdk-for-net#1063]
### Repro
Create a custom model class with a property of non-nullable value type. For example, add a public UnitCount property of type int instead of int?

If you index a document with the default value of that type (e.g. -- 0 for int), the field will be null in Azure Search. If you subsequently search for that document, the Search call will throw JsonSerializationException complaining that it can't convert null to int.

The expected behavior is that `SearchIndexClient` shouldn't allow model classes with non-nullable primitive types. Any call that takes such a type as a parameter should throw ArgumentException if the type contains any primitive public properties.

### Notes
This bug is caused by two things:
1.We don't validate user-defined model classes early.
2.We use DefaultValueHandling.Ignore in JSON.NET to elide nulls in Indexing payloads. This is not a problem per se; We just need to ensure that user-defined model classes don't use non-nullable primitive types.

http://www.newtonsoft.com/json/help/html/SerializationSettings.htm

**Comment by devigned** [comment]
I don't think that we should force custom models to have nullable primitive types. Perhaps zero is a perfectly fine value for a document property. Perhaps zero should be persisted to the data store rather than null.

**Comment by brjohnstmsft** [comment]
To do what you propose would require adding the concept of a default value and non-nullability to document fields in Azure Search. While we could do this in indexing and lookup, we're not able to simulate non-null semantics in filters and faceted searches. This is a consequence of how we store data -- namely, in an inverted index. "Null" is simply how we represent that a document value never appears in the inverted index anywhere.

So the best we could do is allow non-nullable primitive types on the client for indexing data, but disallow them for retrieving data since you can't guarantee only non-null values on the server side. This inconsistency seems worse than disallowing non-nullable primitives, which just ensures that the client model accurately matches the semantics of the server model.

FWIW, no customers have ever asked us for strict non-nullable types. Based on our experience it doesn't appear to be an important requirement for search scenarios.

**Comment by devigned** [comment]
I'm just suggesting you should store the value the user put into the document rather than nulling it out. 

0 == 0 not 0 == null

**Comment by brjohnstmsft** [comment]
Agreed, but that's only going to work with primitive types if we allow them only for document upload, not for retrieval. That seems to defeat the purpose of using a strongly-typed model at all.

**Comment by devigned** [comment]
I might be missing something, but why doesn't it work for retrieval? Could retrieval not send back a 0 as well?

**Comment by brjohnstmsft** [comment]
For that one value, yes. But what about other documents? In general it is very difficult to guarantee that all values for a field will be non-null (e.g. -- you can add new fields to the schema without re-indexing, in which case all values will start out as null). The type system we expose through the Azure Search REST API does not help to enforce non-nullness, nor should it for the efficiency reasons I described above (we wouldn't want to bloat up the inverted index with default values).

**Comment by devigned** [comment]
Would you consider things with a unit count of zero to be bloat to the index? How would you search for things with zero units?

**Comment by brjohnstmsft** [comment]
By "unit count", do you mean something like "units in stock" in a retail scenario? In that case, zero is a semantically important value, so it makes sense to explicitly store it in the index, even if it uses more storage. Putting it in the index ensures that searching for it is fast. Conversely, there are cases like "rating" where an item without a rating is not necessarily better represented with 0 than null. It depends on whether there will be many queries looking for items without a rating. If so, 0 would be better (faster queries, but more space). Otherwise, null would be better (less space, but slower queries).

Bringing this back to the issue at hand... Fixing this zero/null issue *in a robust way* while continuing to allow non-nullable types on the client requires adding support for non-nullable types in the service API. Feel free to propose this on the [Azure Search UserVoice site](http://feedback.azure.com/forums/263029-azure-search) so we can collect more customer feedback to help us prioritize it.

In the meantime, we can fix this scenario on the client by requiring primitive types to be nullable. It may not be a perfect solution, but it's feasible in a much shorter timeframe than changing the service API. If non-nullable types are ever supported in the service API, then easing this restriction on the client should be easy since it would be a backwards-compatible change.

**Comment by SylviaJ** [comment]
I'm sorry, but I don't agree with this comment at all
"In the meantime, we can fix this scenario on the client by requiring primitive types to be nullable."

This is not a solution at all.
Instead for my solution I have had to NOT use this SDK for the uploading of my documents so I can get the correct values into the search index. Instead have had to use the REST API directly and do my own serialization and error handling. What a pain.
if you have ordering of -5, 0, 5, 10 for 4 documents and then the 0 is converted to a null, then that document will not be sorted in the correct order on the search index itself.

I am obviously not part of the coding team here, and the reason why this was added 'DefaultValueHandling.Ignore' I don't really get. I don't see why putting in the actual value of something that you actually want in the index 'bloats' the index. Default values are genuine values and actually need to be in the index.

**Comment by SylviaJ** [comment]
Another scenario where the actual value should be in the index. Say your model has a nullable Boolean. You have 3 possible values, true, false, null. Say you want to filter by null. Because your json serializer is converting all falses to null you are going to get back all of the false matches as well as the null matches. And say that you actually then want to filter by false values. You are going to get back nothing because there are no false values in the search index as they have all been wiped out and set to null.

Default values should never be changed to null. 
>"The expected behavior is that  SearchIndexClient  shouldn't allow model classes with non-nullable primitive types. Any call that takes such a type as a parameter should throw ArgumentException if the type contains any primitive public properties."

This should not be an expected behaviour as having such an expected behaviour means that you end up with incorrect data in your search index.

**Comment by brjohnstmsft** [comment]
I totally agree that the current behavior of serializing default values of non-nullable primitive types (e.g. -- `int`, `bool`) to null is incorrect, and I'm sorry that it has caused you trouble.

With the current release of the SDK, the only way to avoid this behavior is to use nullable primitive types in your model classes (e.g. -- `int?`, `bool?`). There have been several other customers that have run into this problem, but so far switching to nullable types has worked for them. If it looks like using nullable primitives is going to be an unacceptable limitation for many customers, that is good for us to know as it helps us to prioritize a fix. Are you able to share any details about why using nullable types like `int?` and `bool?` wasn't an option in your scenario?

Regarding **DefaultValueHandling.Ignore**: The SDK uses this as a performance optimization for very wide index schemas, where sending thousands of explicit nulls in every indexing batch would result in an unacceptable performance hit. The bug occurs because the SDK assumes it is always serializing nullable types, so default non-nullable values like `0` or `false` are not serialized, causing them to be interpreted as null by the Azure Search service. Neither the SDK nor the Azure Search service were designed to handle non-nullable data types. It's this mismatch between client side non-nullable types and server-side nullable types that is the crux of this issue.

A first cut at a solution would be for the SDK to disallow the use of non-nullable primitive types in model classes. This is something we plan to implement when we take the SDK out of preview some time later this year. In your case it wouldn't have unblocked you, but at least you'd have seen a specific error message to indicate the source of the problem.

A more comprehensive solution would be to actually support non-nullable types in index schemas on the server, but that's a much bigger change. If this is a really important scenario to you, please add an item on our [UserVoice site](http://feedback.azure.com/forums/263029-azure-search) to help us prioritize it.

Thanks and sorry again for the trouble.

-Bruce

**Comment by brjohnstmsft** [comment]
We have a fix ready for this issue. I'll send a PR soon.

After some discussion and investigation, we found a better solution than disallowing non-nullable primitive types, although it involves some small compromises. We're introducing a new `IContractResolver` that looks for non-nullable primitive types and ensures that they use `DefaultValueHandling.Include` instead of `DefaultValueHandling.Ignore`. This means that if you have a model class like this:

    public class Model
    {
        public string Key { get; set; }

        public int IntValue { get; set; }
    }

If you set `IntValue` to 0, that value is now correctly serialized as 0 on the wire and stored as 0 in the index. Round tripping also works great.

There is one potential issue to be aware of with this approach: If you use a model type with a non-nullable property, you have to *guarantee* that no documents in your index contain a null value for the corresponding field. Neither the SDK nor the Azure Search service will help you to enforce this.

This is not just a hypothetical concern: Imagine a scenario where you add a new field to an existing index that is of type `Edm.Int32`. After updating the index definition, all documents will have a null value for that new field (since all types are nullable in Azure Search). If you then use a model class with a non-nullable `int` property for that field, you will get a `JsonSerializationException` like this when trying to retrieve documents:

    Error converting value {null} to type 'System.Int32'. Path 'IntValue'.

For this reason, we still recommend that you use nullable types in your model classes as a best practice.

**Comment by ehuna** [comment]
I'd like to add my 2 cents: I completely with SylviaJ - if my model has 0, I expect a 0 value to be stored and returned, not a null.  This is for sorting purposes, simplifying models and de-serialization, etc...

Since we can't modify fields in indexes, we have created a whole system where new indexes are created when changes are made - so the example above by brjohnstmsft does not apply, we will never run into a scenario where we need to use nullable types.

Thank you for adding DefaultValueHandling.Include !

**Comment by brjohnstmsft** [comment]
FYI, the fix is now available in the latest version of the SDK [on NuGet](http://www.nuget.org/packages/Microsoft.Azure.Search).
